examples/hmm.py

```diff
@@ -0,0 +1,179 @@
+
+    transition_prior = np.ones(num_categories)
+    emission_prior = np.repeat(0.1, num_words)
+
+    transition_prob = dist.Dirichlet(transition_prior).sample(key=rng_transition,
+                                                              size=(num_categories,))
+    emission_prob = dist.Dirichlet(emission_prior).sample(key=rng_emission, size=(num_categories,))
+
+    start_prob = np.repeat(1. / num_categories, num_categories)
+    categories, words = [], []
+    for t in range(num_supervised_data + num_unsupervised_data):
+        rng, rng_transition, rng_emission = random.split(rng, 3)
+        if t == 0 or t == num_supervised_data:
+            category = dist.Categorical(start_prob).sample(key=rng_transition)
+        else:
+            category = dist.Categorical(transition_prob[category]).sample(key=rng_transition)
+        word = dist.Categorical(emission_prob[category]).sample(key=rng_emission)
+        categories.append(category)
+        words.append(word)
+
+    # split into supervised data and unsupervised data
+    categories, words = np.stack(categories), np.stack(words)
+    supervised_categories = categories[:num_supervised_data]
+    supervised_words = words[:num_supervised_data]
+    unsupervised_words = words[num_supervised_data:]
+    return (transition_prior, emission_prior, transition_prob, emission_prob,
+            supervised_categories, supervised_words, unsupervised_words)
+
+
+def forward_one_step(prev_log_prob, curr_word, transition_log_prob, emission_log_prob):
+    log_prob_tmp = np.expand_dims(prev_log_prob, axis=1) + transition_log_prob
+    log_prob = log_prob_tmp + emission_log_prob[:, curr_word]
+    return logsumexp(log_prob, axis=0)
+
+
+def forward_log_prob(init_log_prob, words, transition_lob_prob, emission_log_prob):
+    # Note: The following naive implementation will make it very slow to compile
+    # and do inference. So we use lax.scan instead.
+    #
+    # >>> log_prob = init_log_prob
+    # >>> for word in words:
+    # ...     log_prob = forward_one_step(log_prob, word, transition_log_prob, emission_log_prob)
+    def scan_fn(log_prob, word):
+        return forward_one_step(log_prob, word, transition_lob_prob, emission_log_prob), np.zeros((0,))
+
+    log_prob, _ = lax.scan(scan_fn, init_log_prob, words)
+    return log_prob
+
+
+def semi_supervised_hmm(transition_prior, emission_prior,
+                        supervised_categories, supervised_words,
+                        unsupervised_words):
+    num_categories, num_words = transition_prior.shape[0], emission_prior.shape[0]
+    transition_prob = sample('transition_prob', dist.Dirichlet(
+        np.broadcast_to(transition_prior, (num_categories, num_categories))))
+    emission_prob = sample('emission_prob', dist.Dirichlet(
+        np.broadcast_to(emission_prior, (num_categories, num_words))))
+
+    # models supervised data
+    sample('supervised_categories', dist.Categorical(transition_prob[supervised_categories[:-1]]),
```

Comment:
We are missing the first state. Should we also sample `init_prob` from some prior, and have `sample('init_state', dist.Categorical(init_prob), obs=supervised_categories[0])`?

Reply:
I suppose the current formulation implicitly assumes that the initial state is uniformly distributed? If so, maybe we can just add a comment to that effect?

Reply:
I think that if we don't use `init_prob` to anywhere else in the model then it is not necessary to add that uncertainty into the model. Given an observation categorical[0], we add 3 additional parameters. It is a bit nonsense to me.

Reply:
I don't think that we assume that in this model. Given a data X, y, you can model y ~ SomeDist(X) but need not to model X ~ SomePrior(...).

Reply:
> then it is not necessary to add that uncertainty into the model.

The idea would be to encode our beliefs about the initial state into a prior (and it will likely affect inference depend on how strong this prior is), and get a posterior distribution over `init_prob`. If this prior is flat, then I guess what you are saying is correct, and it makes no difference in inference whether we include it or not (except that we won't end up recording the posterior in that case).

Reply:
> Given a data X, y, you can model y ~ SomeDist(X) but need not to model X ~ SomePrior(...).

Sure, I agree with that. If you don't model it, you are implicitly assuming it to be flat.

Reply:
Yes, you are right. I am assuming it to be flat. I don't know about that fact, so having a comment with `flat prior`, `uniform prior` makes me confused...

Reply:
will a comment "here we don't make any assumption about the first supervised category" make sense to you? It is my language for "here we make an assumption that the first supervised category comes from flat/uniform prior"

Reply:
Sorry, my bad with the sloppy comment. :) Since I suppose we aren't interested in the distribution of the initial state, we can probably just add a comment "Assuming a flat/uniform prior over the simplex for the initial state".

Reply:
Sorry, missed your earlier comment. Sure, feel free to write it in your own words. That makes sense to me!